### PR TITLE
Refactor report

### PR DESCRIPTION
**Pull request by ScottCTD**
- [Rendered Version](https://github.com/UofT-CS-Group/CSC110-Final-Project/blob/scott/Report/final_report.pdf)

### CODE REVIEW

File: Report/final_report.tex [hunk 1]
```diff
@@ -277,33 +293,69 @@
...
-    \subsection{Next steps}
+        \newpage
 
-    We will be carrying the skills we developed in this project into other projects in the future. We have now acquired the ability to download, process, and hash data. These abilities would be necessary to create any automated application in the future.
+        \item The other challenge is about optimizations.
+        
+        In our plot, we implemented a cross-hair that centered on the closet point in the line from the cursor. We successfully implemented it with binary search and many data transformations and calculations. However, the frame rate per second (FPS) was very low when moving the cross-hair.
 
-    We also acquired one of the most important skills in the Computer Science Industry, which is to design a Graphical User Interface. This is the ultimate version of "Designer to User" contract, which provides a very "Easy to use" interface so that users without any knowledge in Computer Science can also use this application to their advantage.
+        To investigate the root of the problem, we analyzed the running time of our algorithms but it turned out that the running time was good ($\bigO(n)$).
```
Comment (Lei-Tin) [inline Report/final_report.tex:340]: > ...we analyzed the running time of our algorithms but it turned out that the running time was good ($\bigO(n)$). 

...we analyzed the running time of our algorithms, but it turned out that the running time was fairly decent ($\bigO(n)$).

File: Report/final_report.tex [hunk 2]
```diff
@@ -277,33 +293,69 @@
...
+        
+        In our plot, we implemented a cross-hair that centered on the closet point in the line from the cursor. We successfully implemented it with binary search and many data transformations and calculations. However, the frame rate per second (FPS) was very low when moving the cross-hair.
 
-    We also acquired one of the most important skills in the Computer Science Industry, which is to design a Graphical User Interface. This is the ultimate version of "Designer to User" contract, which provides a very "Easy to use" interface so that users without any knowledge in Computer Science can also use this application to their advantage.
+        To investigate the root of the problem, we analyzed the running time of our algorithms but it turned out that the running time was good ($\bigO(n)$). 
+        
+        Then, we used \verb|time| module and divided the codes into many piece of codes to calculate the real times each piece took. Finally, we found that the root of the evil is actually the \verb|matplotlib#Canvas.draw()| function call. But here comes the issues, we cannot modify the code from \verb|matplotlib| to make it faster.
```
Comment (Lei-Tin) [inline Report/final_report.tex:342]: > Then, we used \verb|time| module and divided the codes into segments to calculate the real times each piece took. Finally, we found that the root cause of this issue is actually from the \verb|Matplotlib#Canvas.draw()| function call. But another issue arose along with this discovery, we cannot modify the code from \verb|matplotlib| to make it faster.

File: Report/final_report.tex [hunk 3]
```diff
@@ -277,33 +293,69 @@
...
+        
+        After many researches, we found out that we could solve this issue with a relatively advanced method: blit. Since we already had many experiences from the previous challenge, by reading through the documentation, experimenting with many small examples, and updating the whole implementation of \verb|PlotCanvas|, we successfully implement blit to fix the issue.
+        
+        We firstly draw the figure, excluding the cross-hair, and save the figure background to cache it. Then, every time the cursor moves, we only draw the cross-hair and restore the figure from cache. In this way, the previous cross-hair disappears because the figure background does not contain the cross-hair - we draw the figure before the cross-hair, and we don't need to redraw the background frequently.
 
-    This project will be held public on GitHub repository for anyone to fork and create an updated version of the application. We have included docstrings and comments everywhere around our code for anyone that wishes to fork our Repository.
+        This largely improve the FPS and now the cross-hair is very smooth.
```
Comment (Lei-Tin) [inline Report/final_report.tex:348]: > This largely improve the FPS and now the cross-hair is very smooth.

This largely improved the frame rate and now the cross-hair is rendered very smoothly.

File: Report/final_report.tex [hunk 4]
```diff
@@ -277,33 +293,69 @@
...
+    \end{itemize}
+
+    \subsection{The Future}
```
Comment (Lei-Tin) [inline Report/final_report.tex:351]: 我觉得叫 Next Steps 好一点

The future sounds so stupid ngl hahahahahahaha

**Review by Lei-Tin — CHANGES_REQUESTED** [review]
ok 喷的有点多，大部分都是可以直接复制粘贴把我的换上去

你到时候判断觉得我的好还是原本的好，直接选择 Accept Left or Right 就好

File: Report/final_report.tex [hunk 5]
```diff
@@ -89,186 +91,200 @@
...
 
-    For the files that are being downloaded, the application will compare it with the specified MD5 Checksum value we ran on our machines to ensure that the files downloaded are the same as the ones we are working with.
+    \subsubsection{Algorithms}
 
-    In our resource\_manager.py, we have created a class to store all of our different files. These resource objects will contain a remote path, the MD5 identifier, and the local path so we can download and verify the files.
+    To simplify our data manipulations, we abstract out few generic algorithms such as sorting, grouping, searching, etc. in a new module: \verb|algorithms.py|.
 
-    The purpose of this sub-sub-section is to ensure that our application will run on any machine with our predetermined datasets.
```
[the remark this answers is not in the document]
Comment (ScottCTD) [reply Report/final_report.tex:125]: 我这里写的比较泛，没有具体说是sorting algorithm，所以我觉得说compare不太合适

File: Report/final_report.tex [hunk 6]
```diff
@@ -89,186 +91,200 @@
...
-        \item Navigation bar that allows the user to return the plot to the original state
-        \item Scatter plot COVID-19 cases and school closures visualizations supporting filtering by country names
-        \item Customizable color and marker types for plots
-        \item An application with icon
+        \item A fully functional Graphical User Interface with menu bar, multiple buttons, input date edits, sliders, etc.
+        \item A progress bar on the status bar displaying the progress of data initialization.
+        \item Panning and zooming functionalities on the graph by left click, drag, and scrolling.
+        \item COVID-19 cumulative cases scatter plot and school closures visualizations.
+        \item Customizable line color, style, and marker.
+        \item And many more details...
```
[the remark this answers is not in the document]
Comment (ScottCTD) [reply Report/final_report.tex:185]: 啊这 这个就是个总览吧，真要放细节更没必要吧

File: Report/final_report.tex [hunk 7]
```diff
@@ -89,186 +91,200 @@
...
     \end{itemize}
 
-    \subsubsection{Main Window}
+    \subsubsection{Utilities}
```
[the remark this answers is not in the document]
Comment (ScottCTD) [reply Report/final_report.tex:188]: No

Utilities是GUI的Utilities，在这个GUI大类下面，被主GUI用了

File: Report/final_report.tex [hunk 8]
```diff
@@ -89,186 +91,200 @@
...
+        \item \verb|StandardLabel|: A standard text label.
+        \item \verb|StandardPushButton|: A standard button.
+        \item \verb|StandardComboBox|: A standard combo box.
+        \item \verb|StandardDateEdit|: A standard date editor.
+        \item \verb|StandardCheckbox|: A standard checkbox.
+        \item \verb|StandardRadioButton|: A standard radio button.
+        \item \verb|StandardProgress Bar|: A standard progress bar.
+        \item \verb|StandardGroupBox|: A standard group box.
+        \item \verb|StandardMenuBar|: A standard menu bar.
+        \item ...
```
[the remark this answers is not in the document]
Comment (ScottCTD) [reply Report/final_report.tex:204]: 不用

File: Report/final_report.tex [hunk 9]
```diff
@@ -89,186 +91,200 @@
...
-        \item Plot Canvas - The class that is used to plot figures on the Main Window
+        \item \verb|StandardLabel|: A standard text label.
+        \item \verb|StandardPushButton|: A standard button.
+        \item \verb|StandardComboBox|: A standard combo box.
+        \item \verb|StandardDateEdit|: A standard date editor.
+        \item \verb|StandardCheckbox|: A standard checkbox.
+        \item \verb|StandardRadioButton|: A standard radio button.
+        \item \verb|StandardProgress Bar|: A standard progress bar.
+        \item \verb|StandardGroupBox|: A standard group box.
+        \item \verb|StandardMenuBar|: A standard menu bar.
```
[the remark this answers is not in the document]
Comment (ScottCTD) [reply Report/final_report.tex:203]: 嗯。。我稍微加点

File: Report/final_report.tex [hunk 10]
```diff
@@ -277,33 +293,69 @@
...
+        \item We did not introduce the vaccination data and this impacts our scope of analysis. That is, most countries resume in-person learning after most individuals were vaccinated.
+    \end{itemize}
+
+    \subsection{Obstacles}
+
+    \begin{itemize}
+        \item Our application crashes all day, all night, 24 * 7, non-stop. Making us want to drop out of CSC110. But nevertheless, we still overcame the difficulties and produced a "Functioning" application.
+        \item One of the most impressive challenges is the implementation of progress bar. 
+        
+        If we want to display a progress bar monitoring the progress of something like initialization, we need three tasks to be performed simultaneously, or asynchronously:
```
[the remark this answers is not in the document]
Comment (ScottCTD) [reply Report/final_report.tex:315]: 为了引出多线程，我觉得写上好

毕竟是计算机作业，多用点计算机词汇233

File: Report/final_report.tex (hunk 6 above)
[the remark this answers is not in the document]
Comment (Lei-Tin) [reply Report/final_report.tex:185]: No I meant

要不要直接把这行删掉
丢这边好像也没什么用

你可以把前面改成, Functionalities, such as but not limited to:

File: Report/final_report.tex (hunk 8 above)
[the remark this answers is not in the document]
Comment (Lei-Tin) [reply Report/final_report.tex:203]: 那你还不去掉这个"..."

File: Report/final_report.tex (hunk 10 above)
[the remark this answers is not in the document]
Comment (Lei-Tin) [reply Report/final_report.tex:314]: Hmm
因为 Asynchronous 跟 Simulatenous 意思一样，有点 redundant

还有我上面第二个其实是改了语法 + 删掉 Asynchronous

[the remark this answers is not in the document]
Comment (ScottCTD) [reply Report/final_report.tex:314]: ok

File: Report/final_report.tex (hunk 5 above)
[the remark this answers is not in the document]
Comment (Lei-Tin) [reply Report/final_report.tex:125]: ok, make sense
那就留这样吧

File: Report/final_report.tex (hunk 7 above)
[the remark this answers is not in the document]
Comment (Lei-Tin) [reply Report/final_report.tex:187]: oh shit I'm blind
I thought it was Utilities as an individual section
没看到 subsub section，那没事了

**Review by Lei-Tin — APPROVED** [review]
Okai，再看了一遍没看到什么大问题
我明天无聊了，有空了再瞅瞅 :3

就是 for some, idk why you marked discussion as resolve but you didn't commit a change to fix it
目前只发现了[这个](https://github.com/UofT-CS-Group/CSC110-Final-Project/pull/44/files#diff-3250e24ae1ef9b1678812561f78c7ea59003b34e33310a3a026765af20a45cd6R202)

有请我们的下一位 Reviewer！@CharlotteChenyy 
我先去躺平了哈哈哈哈哈哈哈哈哈

File: Report/final_report.tex [hunk 11]
```diff
@@ -277,43 +291,79 @@
...
+        \item A major limitation of our application is the fact that our data is not a 100\% accurate data from all around the globe. School Closure Data being used is uploaded by an individual. The COVID-19 Data may also contain inaccurate information.
+        \item Furthermore, we recklessly removed some of the country names in our datasets because they contain non-ASCII characters, which limits the amount of data.
+        \item We also removed some country names because they were not present in both datasets, making the actual data presented in our application even more limited.
+        \item We did not introduce the vaccination data and this impacts our scope of analysis. That is, most countries resume in-person learning after most individuals were vaccinated.
+    \end{itemize}
+
+    \subsection{Obstacles}
+
+    \begin{itemize}
+        \item Our application crashes all day, all night, 24 * 7, non-stop. Making us want to drop out of CSC110. But nevertheless, we still overcame the difficulties and produced a "Functioning" application.
```
Comment (Lei-Tin) [inline Report/final_report.tex:310]: btw 这个哈哈哈哈哈哈哈哈哈啊哈哈
看看 Charlotte 跟 Alysssa 要不要留吧哈哈哈哈哈哈哈哈哈

超级 informal but pretty fun 哈哈哈哈哈

File: Report/final_report.tex [hunk 12]
```diff
@@ -89,186 +91,198 @@
...
+        \item It can run on a M1 Mac with an extensive setup. (You need to run Terminal on Rosetta and install the x86\_64 version of all libraries used through pip)
+        \item We are unsure of the performance of our application in an Intel Mac.
+        \item We haven't tested our application on any Linux distributions.
     \end{itemize}
 
+    \subsection{Remedy}
+
+    If our application failed to download all \textbf{required} resources, please download this compressed file and unzip it in the same directory as the \verb|main.py|. Make sure that the \verb|resources| folder is in the same directory as the \verb|main.py|. TODO: Add the link before we submit.
+
+    Or you could clone our repository and everything will be right in the correct place. TODO: Add github.
```
Comment (Lei-Tin) [inline Report/final_report.tex:275]: GitHub link: 
https://github.com/UofT-CS-Group/CSC110-Final-Project

ssh: 
git@github.com:UofT-CS-Group/CSC110-Final-Project.git

idk 你要放哪个, 如果是link你应该已经放了，so 很有可能是我不知道我自己在干嘛

File: Report/final_report.tex [hunk 13]
```diff
@@ -89,186 +91,198 @@
...
-        \item We are unsure of the performance of our application in an Intel Mac
+        \item We only tested our application in Windows 10/11 platform with Python 3.10.1. Although we were trying our best to make it compatible with others, we could not guarantee the behaviors under other operating systems.
+        \item It can run on a M1 Mac with an extensive setup. (You need to run Terminal on Rosetta and install the x86\_64 version of all libraries used through pip)
+        \item We are unsure of the performance of our application in an Intel Mac.
+        \item We haven't tested our application on any Linux distributions.
     \end{itemize}
 
+    \subsection{Remedy}
+
+    If our application failed to download all \textbf{required} resources, please download this compressed file and unzip it in the same directory as the \verb|main.py|. Make sure that the \verb|resources| folder is in the same directory as the \verb|main.py|. TODO: Add the link before we submit.
```
Comment (Lei-Tin) [inline Report/final_report.tex:273]: Maybe use [USend](https://send.utoronto.ca/)

This is the one recommended in the submission webpage

Comment (Lei-Tin) [reply Report/final_report.tex:273]: This one is available for 14 days, maybe we can upload it on Dec 12 or 13

File: Report/final_report.tex (hunk 8 above)
[the remark this answers is not in the document]
Comment (ScottCTD) [reply Report/final_report.tex:202]: 我的意思是留着，但是不用补充其他的，因为确实还有别的

File: Report/final_report.tex (hunk 11 above)
Comment (ScottCTD) [reply Report/final_report.tex:310]: 留

File: Report/final_report.tex (hunk 8 above)
[the remark this answers is not in the document]
Comment (Lei-Tin) [reply Report/final_report.tex:202]: Hmm, 我觉得这种列举但是不全说的在最后加个 "And more" 听起来就不太对

你可以参考[上面](https://github.com/UofT-CS-Group/CSC110-Final-Project/pull/44/files#diff-3250e24ae1ef9b1678812561f78c7ea59003b34e33310a3a026765af20a45cd6R177)那个做法，改成 "the following classes, such as but not limited to: "

File: Report/final_report.tex (hunk 13 above)
Comment (ScottCTD) [reply Report/final_report.tex:274]: Google Drive public link

No time limit if I were not wrong

Comment (Lei-Tin) [reply Report/final_report.tex:274]: That would work too, 但是要确保有让任何人都能打开的 权限

File: Report/final_report.tex [hunk 14]
```diff
@@ -174,15 +174,14 @@
...
 
-    We implemented the following functionalities:
+    We implemented the following functionalities, such as but not limited to:
 
     \begin{itemize}
         \item A fully functional Graphical User Interface with menu bar, multiple buttons, input date edits, sliders, etc.
         \item A progress bar on the status bar displaying the progress of data initialization.
         \item Panning and zooming functionalities on the graph by left click, drag, and scrolling.
         \item COVID-19 cumulative cases scatter plot and school closures visualizations.
         \item Customizable line color, style, and marker.
```
Comment (CharlotteChenyy) [inline Report/final_report.tex:184]: Maybe add: Some shortcuts of actions

File: Report/final_report.tex [hunk 15]
```diff
@@ -89,186 +91,200 @@
...
-            Either use PyCharm or a terminal to run "main.py" and you should be able to see a window popup
+            Either use PyCharm or a terminal to run "main.py" and you should be able to see a window popup.
 
         \item [5]
-            Select a "Sorting algorithm" and click initialize. After a short amount of time, the application should be ready
+            Select a "Sorting algorithm" and click the initialize button. After few seconds, our application should be ready.
 
         \item [6]
-            You should be able to set start and end dates for the plot, pan and zoom on the plot, select countries and plot the result again, and utilize the menu bar for more customizable configurations.
+            You should be able to set the start and end dates for the plot, pan and zoom on the plot, select countries and plot the result again, and utilize the menu bar for more customizable configurations.
```
Comment (CharlotteChenyy) [inline Report/final_report.tex:256]: > ...and utilize the menu bar for more customizable configurations.

...and utilize the menu bar for more customizable configurations or even save the plot.

File: Report/final_report.tex [hunk 16]
```diff
@@ -89,186 +91,200 @@
...
+    \subsection{Notes}
 
     \begin{itemize}
-        \item The application we've created is only fully compatible with the Windows operating system.
-        \item It can run on a M1 Mac with very extensive setup, but it will be too difficult to set it up. (You need to run Terminal on Rosetta and install the x86\_64 version of all libraries used through pip)
-        \item We are unsure of the performance of our application in an Intel Mac
+        \item We only tested our application in Windows 10/11 platform with Python 3.10.1. Although we were trying our best to make it compatible with others, we could not guarantee the behaviors under other operating systems.
+        \item It can run on a M1 Mac with an extensive setup. (You need to run Terminal on Rosetta and install the x86\_64 version of all libraries used through pip)
+        \item We are unsure of the performance of our application in an Intel Mac.
+        \item We haven't tested our application on any Linux distributions.
```
Comment (CharlotteChenyy) [inline Report/final_report.tex:269]: 我今天可以晚些时候在虚拟机上的linux系统跑一下 然后再来comment

File: Report/final_report.tex (hunk 11 above)
Comment (CharlotteChenyy) [reply Report/final_report.tex:311]: lololol

**Review by CharlotteChenyy — COMMENTED** [review]
我comment的可加可不加 我不知道合不合适lol

File: Report/final_report.tex (hunk 11 above)
Comment (CharlotteChenyy) [reply Report/final_report.tex:311]: lololol

File: Report/final_report.tex (hunk 16 above)
Comment (ScottCTD) [reply Report/final_report.tex:269]: 2333我在WSL跑不起来，不过用虚拟机估计可以

**Comment by CharlotteChenyy** [comment]
image
linux似乎跑不起来 然后我在另一个windows里跑也是同样的报错信息

**Comment by CharlotteChenyy** [comment]
咱们还考虑多平台吗 感觉不满足python3.6+的兼容性是不是就别考虑多平台了 很多发行版源还是3.7 3.8的 自行编译试过了 x-server会崩

**Comment by ScottCTD** [comment]
我不想考虑 现在最正常的是windows + python3.10

直接report里说我们只这样测试过就行

**Comment by Lei-Tin** [comment]
> 我不想考虑 现在最正常的是windows + python3.10
> 
> 直接report里说我们只这样测试过就行

Ye，可以的

实际情况就是 Win 10 + Python 3.10 (最稳定)
Or MacOS Monterey + Intel + Python 3.10
Or MacOS Big Sur + M1 + Rosetta Terminal + Python 3.9.6 + A ton of compatibility configurations

**Review by CharlotteChenyy — APPROVED** [review]

**Comment by ScottCTD** [comment]
ok 都approve，merge里，有问题直接新pr就行

**Pull request merged**